### PR TITLE
Feature/email change

### PR DESCRIPTION
**Pull request by niteshgrg**
Allow user to change his email.

Open To-Do:
- [x] add a warning before changing email / logging out the user?
- [x] add missing translation in backend https://github.com/Plant-for-the-Planet-org/treecounter-app/pull/715#issuecomment-536466176 - @jmiridis can you help here? -> added own issue #1559

### CODE REVIEW

**Comment by norbertschuler** [comment]
This is pretty old now. Will this ever be used any more? If not I suggest should close this pull request. (We can keep the branch if the code is valuable.)

**Comment by norbertschuler** [comment]
Nitesh wrote:
715 is email change functionality which should work as it is. You can check it out and merge it.
It got left out when it was first created at that time we were going into public release so didn’t wanted to risk merging any new code at that point.

**Comment by norbertschuler** [comment]
This patch works fine, but I won't merge it as the backend does not really seem to change the email after you click the link in the confirmation mail received. So it does not make sense to add this functionality in the frontend before the backend is ready for it.
@jmiridis am I right that email change isn't working in the backend yet?

**Comment by sagararyal** [comment]
@norbertschuler Changing Email in backend does not require email confirmation and work fine, so it could simply be issues with APIs.

**Comment by norbertschuler** [comment]
> @norbertschuler Changing Email in backend does not require email confirmation and work fine, so it could simply be issues with APIs.

Yeah, right, that's what I meant. The API call of the backend does not yet work.

**Comment by jmiridis** [comment]
Testing profile change via Postman shows that everything works fine. Here the protocoll:

#### UserProfile DB record
```
"id"	"email"	"new_email"	"activation_token"
"21327"	"jmiridis@gmail.com"
```

#### Request
```
curl -X PUT \
  https://test.trilliontreecampaign.org/api/v1.1/en/profileEmailUpdate \
  -H 'Content-Type: application/json' \
  -d '{
    "newEmail": {
      "first": "jmiridis+2@gmail.com",
      "second": "jmiridis+2@gmail.com"
    }
}'
```

#### Response:
```
{
    "id": 21327,
    "newEmail": "jmiridis+2@gmail.com"
}
```

#### Email
```
#### UserProfile DB record
```
"id"	"email"			"new_email"		"activation_token"
"21327"	"jmiridis@gmail.com"	"jmiridis+2@gmail.com"	"84b85eb5d711426559de4b6ac482ab80"
```

#### Activation Request
```
curl -X POST \
  https://test.trilliontreecampaign.org/auth/v1.1/en/accountActivate \
  -H 'Content-Type: application/json' \
  -d '{
  "token": "84b85eb5d711426559de4b6ac482ab80"
}'
```

#### Response:
```
{
    "token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJpYXQiOjE1Njk4MDc1NzgsImV4cCI6MTU2OTgxMTE3OCwicm9sZXMiOlsiUk9MRV9UUkVFRE9OT1IiLCJST0xFX0JBQ0tFTkQiLCJST0xFX0FMTE9XRURfVE9fU1dJVENIIiwiUk9MRV9SRUNPVkVSWSIsIlJPTEVfVVNFUiJdLCJ1c2VybmFtZSI6ImptaXJpZGlzKzJAZ21haWwuY29tIn0.l50nNWAmBEbKqBghlorMFG4X_MR4DuBaGR9-ZUx6WKfsrHePmOd1kh4O3nh4GuJrcvfFw7Hpf7Y5h1t-MOkhgZMvFB6zkC2wbeyPuCzsj31W87e2_FT6xdxn-iPTy9njf3HtICu6yGaSJoRGduoaa4n6HYpagEUsMqqr3wWmFvmryUA8JnG09mIgGdpV0KcfS0A8_BmnGQmpqitfDCcG3_2KvDBLn1jm6NkSpxEYW-D90wkDB2ZsMeVWMMsBnXf9W6EXJ0Avn4ODVNkEbZdMef4XjOrodN4_gQztlK3vwirehDeRrK8iz2cuSa4jtQXb_HFUqO8gMW4rNOns4KUPO03SDSozr3bQ5Av001KImUfb9ZJH8SZchS4H_RJHmdxlmMS7_YV4iroRg-H_LIhphUyWsF_TiTCCfOYWPVUjidwm-IAtwJN6AcGu_ve-uKZj0R1XJq7yzo_lOuW9VTotbjVdFPmiVfAEEUid97ep55QZrKczaqvYGPjwAMo1F6yLz_DOr-iKBI4K-Wvdk54lV_L6h86oX6wMaCIu45lE8uJms7x-rb5C3kiqA9dIIHhBLWAonQwuUn7xbwBFPnrrzGrA7qMfE-QIS2nGSega8sP5MSw5dxuaKti8nk8d-nqAWZi6o4QZtfPk52IWB7u_7oQ2v5hJioGt5421WYz1OXk",
    "refresh_token": "9ed4df09f74c32727ce7f7ceb8b7634c08c365c7e2a980e88b95809faa603c1b71347f926c322e6c6f92049dcc2b29796a9bfee5845eafb0b08f9e3c40ba39fd",
    "data": {
        "id": 21327,
        "isActivated": true,
        "type": "individual",
        "routeName": "app_userHome",
        "routeParams": [],
        "roles": [
            "ROLE_TREEDONOR",
            "ROLE_BACKEND",
            "ROLE_ALLOWED_TO_SWITCH",
            "ROLE_RECOVERY",
            "ROLE_USER"
        ],
        "maySwitchUser": true,
        "updatedEmail": "jmiridis+2@gmail.com"
    }
}
```

#### UserProfile DB record
```
"id"	"email"			"new_email"	"activation_token"
"21327"	"jmiridis+2@gmail.com"		
```

**Comment by jmiridis** [comment]
@norbertschuler can you have a look why this seems to not work on the client side?

**Comment by norbertschuler** [comment]
@jmiridis yes, it works on test, but not yet on production - isn't is released yet? On production the last step clicking on the link in the confirmation email does not change the email of my account data.

**Comment by norbertschuler** [comment]
I found a missing translation if you enter a email already existing in the database:
```
{
  "code": 400,
  "message": "Validation Failed",
  "errors": {
    "children": {
      "newEmail": {
        "errors": [
          "someone else is already using this email address"
        ],
        "children": {
          "first": {
          },
          "second": {
          }
        }
      }
    }
  }
}
```

**Comment by norbertschuler** [comment]
@jmiridis it does not yet work for me on production, so I can not merge this. Can you help me verify if it should be working? Eventually it only works once, but not twice (try to change the mail address back again).

**Comment by norbertschuler** [comment]
From @jmiridis - this is the link the user clicks on:
```
/account-activate/34e50e9419ffa5be8a633e024dc13433  (app_accountActivate)
```
The application must then call:
```
/auth/v1.x/en/accountActivate (auth_accountActivate_post)
```
posting the token
```
{
  "token": "34e50e9419ffa5be8a633e024dc13433"
}
```
Instead it is calling again:
```
/api/v1.1/en/profileEmailUpdate
```

**Comment by norbertschuler** [comment]
This process works very well for an logged out user as the link https://test.trilliontreecampaign.org/account-activate/XXXX is the same as the link we use for verifying new registrations, but when the user is still logged in, the link does not work as the user is redirected to the home page without triggering the API call @jmiridis has described above.

So we either need to logout the user after we has changed his email address and change the account activation process to also work for logged in users.

**Comment by norbertschuler** [comment]
I decide we just logout the user when we changed his email, so he has to wait for the email to activate the account again.

File: app/containers/EditUserProfile/index.js [hunk 1]
```diff
@@ -213,6 +214,7 @@ class EditUserProfileContainer extends React.Component {
         openPasswordUpdatedDialog={this.state.showPasswordDialog}
         handlePaswordUpdatedClose={this.handleCloseModal}
         deleteProfile={this.deleteProfile}
+        updateEmail={this.props.updateEmail}
```
Comment (norbertschuler) [inline app/containers/EditUserProfile/index.js]: I wonder if we could move some code from app/actions/updateUserProfile.js into this component to make it similar to `this.deleteProfile`.

Comment (mahmed0715) [reply app/containers/EditUserProfile/index.js]: We need to see if that method is being passed to other components, if not we can move that to that inner component.
Agreed with logging out user when they change their email. We must show a notification(get confirmation) before we send request to backend,

Comment (norbertschuler) [reply app/containers/EditUserProfile/index.js]: I already refactored it now. We already gets a note that he will be sent an email. But showing the user a warning before logging him out would be of course another thing to do. But that's beyond my react-native skills yet.

**Comment by mahmed0715** [comment]
@norbertschuler please create an issue specifying the condition, Me or someone else will take care of that...

File: app/containers/EditUserProfile/index.js (hunk 1 above)
Comment (norbertschuler) [reply app/containers/EditUserProfile/index.js]: I added a note to the form about being logged out:

Bildschirmfoto 2019-10-22 um 09 41 41

**Comment by norbertschuler** [comment]
> @norbertschuler please create an issue specifying the condition, Me or someone else will take care of that...

Thank you, but I solved it myself now by adding a warning in the page.

**Pull request merged**